### DISCUSSION TITLE
OptionList with only one parameter does not return highlighted

### DISCUSSION THREAD

**Discussion by CognitiveDisson** [Textualize/textual#2857] [Q&A] [answered]
I subclassed OptionList, added an on_key callback, and noticed that if the list contains only one option, the highlighted property always None.
I subclassed OptionList, added an on_key callback, and noticed that if the list contains only one option, the highlighted property always None. (even when you select the row with that option and it's visually highlighted)

```
class ListWidget(OptionList):
    def on_mount(self) -> None:
        self.add_option(Option("1"))

    def on_key(self, event: events.Key) -> None:
        print(f"highlighted: {self.highlighted}")

```
In `textual console`:
```
highlighted: None
```

**Comment by davep** [comment]
I'd expect this behaviour, which doesn't have anything to do with just having one option. If you create an `OptionList` with no options, and then add more afterwards, nothing with be highlighted until some sort of input causes an option to be highlighted. As such `highlighted` should be `None`.

See this in action with this little app, which creates an `OptionList` with no options, adds some `on_mount`, logs `highlighted`, and then further logs `highlighted` when the highlight is moved.

```python
from textual import on
from textual.app import App, ComposeResult
from textual.containers import Vertical
from textual.widgets import OptionList, TextLog
from textual.widgets.option_list import Option

class OptionListApp(App[None]):

    CSS = """
    OptionList {
        height: 4fr;
    }

    TextLog {
        height: 1fr;
    }
    """

    def compose(self) -> ComposeResult:
        with Vertical():
            yield OptionList()
            yield TextLog()

    def on_mount(self):
        for n in range(20):
            self.query_one(OptionList).add_option(Option(f"Here is option {n}"))
        self.log_highlighted()

    @on(OptionList.OptionHighlighted)
    def log_highlighted(self) -> None:
        highlighted = self.query_one(OptionList).highlighted
        self.query_one(TextLog).write(f"Highlighted: {highlighted!r}")

if __name__ == "__main__":
    OptionListApp().run()
```

**Comment by nmay231** [comment]
> I subclassed OptionList, added an on_key callback, and noticed that if the list contains only one option, the highlighted property always None. (even when you select the row with that option and it's visually highlighted)

@CognitiveDisson As already mentioned, if there is no option selected, it *should* print `None`. However, I got `highlighted: 0` when I selected the row, so there must be something else going on for you to get `None` even after the row is selected.

**Comment by willmcgugan** [comment, marked as answer]
The issue was that the OptionList didn't have an opertunity to process the key bindings. Responding to the `OptionList.OptionHighlighted` event as @davep suggested will ensure that your code runs when the highlight changes.

This code reproduces the issue (and includes the solution).

```python
from textual.app import App, ComposeResult
from textual.widgets import OptionList
from textual.widgets.option_list import Option
from textual import events
from textual import on


class ListWidget(OptionList):
    def on_mount(self) -> None:
        self.add_option(Option("1"))

    def on_key(self, event: events.Key) -> None:
        print(f"highlighted: {self.highlighted}")

    @on(OptionList.OptionHighlighted)
    def highlight_changed(self, event: OptionList.OptionHighlighted) -> None:
        print(f"highlighted (change): {self.highlighted}")


class TestApp(App):
    def compose(self) -> ComposeResult:
        yield ListWidget()


app = TestApp()
app.run()
```

**Comment by CognitiveDisson** [comment]
Thanks for helping solve that issue. The suggested solution works for me.
